tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -558,3 +584,114 @@ def prepare_first_guess(
             index += atomic_number + 1
         first_guess[-1] = electron_density
         return first_guess
+
+    def population_objective_function(
+        self,
+        populations,
+        atomic_numbers,
+        phi,
+        solution_vector,
+        rate_matrix_index,
+        total_photo_ion_coefficients,
+        total_rad_recomb_coefficients,
+        total_coll_ion_coefficients,
+        total_coll_recomb_coefficients,
+    ):
+        """Main set of equations for the NLTE ionization solver. A*x - B = 0, where x is the populations,
+        A is the matrix of rates and B is the solution vector.
+
+        Parameters
+        ----------
+        populations : numpy.array
+            Current values of ion number densities and electron density.
+        atomic_numbers : numpy.array
+            All atomic numbers present in the plasma.
+        phi :DataFrame
```

Comment:
```suggestion
        phi : DataFrame
```

Reply:
I would say `pandas.DataFrame` it might not understand otherwise